### ISSUE TITLE
Encoding runs in worst-case quadratic time

### ISSUE THREAD

**Issue by appgurueu** [rxi/json.lua#45]
Consider the following benchmark which encodes the deeply nested structure `[1000, [999, [998, [...]]]]`:

```lua
local json = require"json"
for _ = 1, 1e3 do
	local t = {}
	for i = 1, 1e3 do
		t = {i, t}
	end
	json.encode(t)
end
```

this should run in a fraction of a second. Instead, it takes ~2 seconds on my device. This is because string concatenations / recursive `table.concat` are internally used to encode the JSON:

```lua
return "[" .. table.concat(res, ",") .. "]"
```

This means that the outer loop will have to add the `[` and `]` brackets in time linear in `table.concat(res, ",")`, which in our example will be just marginally shorter, which in turn will only differ in length by a constant. Thus you do a linear time operation linearly often, resulting in worst-case **quadratic encoding runtime** in the depth of the nested structure.

The proper fix is to use a shared rope and have all recursive calls write to the same rope. You then only concatenate strings once at the end. Building the string then runs in linear time. You can find an example of this approach in [my JSON implementation](https://github.com/appgurueu/modlib/blob/72ca08ed3a955f614fb853e37f7e31a5f5ec2999/json.lua#L290).

**Comment by flamendless** [comment]
Perhaps you could make a PR implementing that solution :)

**Comment by vurvdev** [comment]
If you want a fast implementation: https://github.com/Vurv78/qjson.lua

Runs thousands of times faster than this repo

**Comment by appgurueu** [comment]
> If you want a fast implementation: https://github.com/Vurv78/qjson.lua
> 
> Runs thousands of times faster than this repo

Refer to [this comment by me](https://github.com/rxi/json.lua/issues/34#issuecomment-1613136352).

**TL;DR:** Your JSON parser is **broken** and has a **time complexity issue** as well. Your performance claims in their current form can't be taken seriously.

**Comment by vurvdev** [comment]
> > If you want a fast implementation: https://github.com/Vurv78/qjson.lua
> > Runs thousands of times faster than this repo
> 
> Refer to [this comment by me](https://github.com/rxi/json.lua/issues/34#issuecomment-1613136352).
> 
> **TL;DR:** Your JSON parser is **broken** and has a **time complexity issue** as well. Your performance claims in their current form can't be taken seriously.

Both of these have been addressed. Thanks for bringing it to my attention (also see my other comment)
